### DISCUSSION TITLE
React-hook-form: Type 'T' is not assignable to type 'Path<TFieldValues>'

### DISCUSSION THREAD

**Discussion by perevezencev** [react-hook-form/react-hook-form#11680] [Q&A]
Hello everyone! I have a code like this:

```typescript
import React from "react"
import { z } from "zod"
import { Control, Controller } from "react-hook-form" 

export const address = z.object({
    street: z.string(),
})

type FieldValues<TPrefix extends string> = {
    addresses: {
        [k in TPrefix]: z.infer<typeof address>
    }
}

export type AddressProperties<
    TFieldValues extends FieldValues<TPrefix>,
    TPrefix extends string = string,
> = {
    control: Control<TFieldValues>
    prefix: TPrefix
}

export function Address<TFieldValues extends FieldValues<TPrefix>, TPrefix extends string>(
    properties: AddressProperties<TFieldValues>,
) {
    const { control, prefix} = properties

    return <Controller 
        control={control} 
        // Type '`addresses.${string}.street`' is not assignable to type 'Path<TFieldValues>'
        name={`addresses.${prefix}.street`}
        render={() => <input />} 
     />
}
```

why `name` is not assignable to type `Path<TFieldValues>` and how can i fix this?

[playground][1]

  [1]: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChRJY4BvOALzgF85tc4AiOiAE3fPGjxaAYQgA7GDgA2AGjiiJ0qUihMWOPOyLoYAWgAWECAGtdmaCHZxSpJAA9K8NOIDO8FN25EXLuAF56ADoIACMAKyQMAApqUjh4uDciJBgALiCk4DEAcyiAShlSRjybGABPMCQ4ADFgJCluADUUKQBXJBcAHgAVAAUiTGA7OHsYJDFuX0ycgD5-GjiEjy8Olw702IStuABtYzgsuD6BoYBddLpArMwVTvLKiEw4Ze8XGcX4xiKbe0c4e6qAEFPK9%20hBKrA6l0PkdavUmi12r5RuNJjU6g1mm0Oj1%20khBnYZoUtsd8UMRnYxhMppIstl5tNsoU5gFNglnIoIFJ0gpJFyenDMYiOu8tmATnZ0qSCd9bA5BCxWmIMMBxHBgSsfAKMQjscjKajfILdUjcRKiUc8QSKVS0YyZlEYeLwSoYFD0hrQTgIW6cd1jVikUTSHkFlsOW4aHAOXzZHBxWS7MwAs6fVCbFsiDBWlAxHBOrylCprNt2eJY35qDHpMwYVsAPT1o4VKr4AAGL1WHUCABJqIzGIEkkgUm38AdfGIIO4fMBsmIUCFlP8IP8WwReigYPptfDAyKyKX4guQEhKx2QV2XL3qAmCYPh6Ovke4EQJipK-l-HNOlkwK14HrGZa22IDviAA

**Comment by sujinleeme** [comment]
Hi, @perevezencev  You'd like to give the generic type of form value object. The type definition is difficult to read and is complicated. We can make it simpler.  
 
```ts
import React from 'react'
import { z } from "zod"
import { Control, Controller } from "react-hook-form" 

export const address = z.object({
    street: z.string(),
})

type FormValues<Prefix> = {
    [K in keyof Prefix | string]: z.infer<typeof address>
}

type AddressProperties<Prefix> = {
    control: Control<FormValues<Prefix>>, 
    prefix: Prefix
}

export function Address(
    properties: AddressProperties<'address-1' | 'address-2'>
) {
    const { control, prefix} = properties

    return <Controller 
        control={control} 
        name={`addresses.${prefix}.street`}
        render={() => <input />} 
     />
}
```

[Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChRJY4BvOALzgF85tc4AiOiAE3fPGjxaAYQgA7GDgA2AGjiiJ0qUihMWOPOyLoYAWgAWECAGtdmaCHZxSpJAA9K8NOIDO8FN25EXLuAF56ADoIACMAKyQMAApqUjh4uDciJBgALiCk4DEAcyiAShlSRjzSGABPMCQ4ADELADUUKQBXJBcAHgAFIkxgOwA+fxo4hIBtAGk4LLhjJDKITDgupB67OAAfRMks7IBddLpArMwVNvLK+bgPL1aXPqKbM6qAQU9vFy6ISthgVs7u3oGAViCTgzkUECk6QUkghbVqUBADWavyWKz6fTkw3iYH+dnSqN691sDkELCaYgwwHEcBe1x8USxcBxnxUMB+LnStLeHy+bN++Cub10AEZ8OsCILWroAEz4O55IYgsFuGig8Qw2RM3HMALM3nsmwgogwJpQMRwNrQpQqawgpXq6R+ahgjXMRkgsQoEBIJ0AA0lPlagQAJNQccteoxAkkkClfYx3QkiGJuConfl-AM2lkwE14AB6Ppuu2F+5AA)
